pkg/controller/hvpa/hvpa_controller.go

```diff
@@ -840,19 +892,68 @@ func getWeightedReplicas(hpaStatus *autoscaling.HorizontalPodAutoscalerStatus, h
 	scaleUpStabilizationWindow, _ := time.ParseDuration(*hvpa.Spec.ScaleUpStabilization.Duration)
 	scaleDownStabilizationWindow, _ := time.ParseDuration(*hvpa.Spec.ScaleDownStabilization.Duration)
 
-	if weightedReplicas > currentReplicas && (overrideScaleUpStabilization || lastScaleTimeDuration > scaleUpStabilizationWindow) {
-		log.V(2).Info("HPA scaling up", "weighted replicas", weightedReplicas)
-		return weightedReplicas, err
-	} else if weightedReplicas < currentReplicas && lastScaleTimeDuration > scaleDownStabilizationWindow {
-		log.V(2).Info("HPA scaling down", "weighted replicas", weightedReplicas)
-		return weightedReplicas, err
+	updateMode := *hvpa.Spec.Hpa.UpdatePolicy.UpdateMode
+
+	if hpaWeight == 0 {
+		blockReason = autoscalingv1alpha1.BlockingReasonWeight
+
+	} else if updateMode == autoscalingv1alpha1.UpdateModePurge || updateMode == autoscalingv1alpha1.UpdateModeOff {
+		blockReason = autoscalingv1alpha1.BlockingReasonUpdatePolicy
+
+	} else if weightedReplicas > currentReplicas {
+		if overrideScaleUpStabilization == false && lastScaleTimeDuration < scaleUpStabilizationWindow {
+			blockReason = autoscalingv1alpha1.BlockingReasonStabilizationWindow
+		} else {
+			log.V(2).Info("HPA scaling up", "weighted replicas", weightedReplicas)
+			outHpaStatus.DesiredReplicas = weightedReplicas
+			return outHpaStatus, err
```

Comment:
We should return here only if `weightedReplicas == desiredReplicas`. If not, we should add the `blockedScaling` with reason `BlockingReasonWeight` even if `weightedReplicas != currentReplicas`.

But we can pick this as a separate issue.

Reply:
Resolved here in favour of #20